ospfd/ospf_sr.c

```diff
@@ -0,0 +1,2190 @@
+
+	if (size > MPLS_DEFAULT_MAX_SRGB_SIZE || size <= 0) {
+		vty_out(vty,
+			"Range size cannot be less than 0 or more than %d\n",
+			MPLS_DEFAULT_MAX_SRGB_SIZE);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	if (upper > MPLS_DEFAULT_MAX_SRGB_LABEL) {
+		vty_out(vty, "Upper-bound cannot exceed %d\n",
+			MPLS_DEFAULT_MAX_SRGB_LABEL);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	if (upper < MPLS_DEFAULT_MIN_SRGB_LABEL) {
+		vty_out(vty, "Upper-bound cannot be lower than %d\n",
+			MPLS_DEFAULT_MIN_SRGB_LABEL);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	/* Set SID/Label range SRGB */
+	OspfSR.srgb.range_size = size;
+	OspfSR.srgb.lower_bound = lower;
+
+	/* Set Router Information SR parameters */
+	ospf_router_info_update_sr(true, OspfSR.srgb, OspfSR.msd);
+
+	/* Update NHLFE entries */
+	hash_iterate(OspfSR.neighbors,
+		     (void (*)(struct hash_backet *, void *))update_in_nhlfe,
+		     NULL);
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (no_sr_sid_label_range,
+	no_sr_sid_label_range_cmd,
+	"no segment-routing global-block",
+	NO_STR
+	SR_STR
+	"Delete Segment Routing Global Block label range\n")
+{
+
+	if (!ospf_sr_enabled(vty))
+		return CMD_WARNING_CONFIG_FAILED;
+
+	/* Revert to default SRGB value */
+	OspfSR.srgb.range_size = MPLS_DEFAULT_MIN_SRGB_SIZE;
+	OspfSR.srgb.lower_bound = MPLS_DEFAULT_MIN_SRGB_LABEL;
+
+	/* Set Router Information SR parameters */
+	ospf_router_info_update_sr(true, OspfSR.srgb, OspfSR.msd);
+
+	/* Update NHLFE entries */
+	hash_iterate(OspfSR.neighbors,
+		     (void (*)(struct hash_backet *, void *))update_in_nhlfe,
+		     NULL);
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (sr_node_msd,
+       sr_node_msd_cmd,
+       "segment-routing node-msd (1-16)",
+       SR_STR
+       "Maximum Stack Depth for this router\n"
+       "Maximum number of label that could be stack (1-16)\n")
+{
+	uint32_t msd;
+	int idx_number = 2;
+
+	if (!ospf_sr_enabled(vty))
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (sscanf(argv[idx_number]->arg, "%d", &msd) != 1) {
+		vty_out(vty, "segment-routing: fscanf: %s\n",
+			safe_strerror(errno));
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	if (msd < 1 || msd > MPLS_MAX_LABELS) {
+		vty_out(vty, "MSD must be comprise between 1 and %d\n",
+			MPLS_MAX_LABELS);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	/* Set this router MSD */
+	OspfSR.msd = msd;
+
+	/* Set Router Information SR parameters */
+	ospf_router_info_update_sr(true, OspfSR.srgb, OspfSR.msd);
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (no_sr_node_msd,
+	no_sr_node_msd_cmd,
+	"no segment-routing node-msd",
+	NO_STR
+	SR_STR
+	"Disable Maximum Stack Depth for this router\n")
+{
+
+	if (!ospf_sr_enabled(vty))
+		return CMD_WARNING_CONFIG_FAILED;
+
+	/* unset this router MSD */
+	OspfSR.msd = 0;
+
+	/* Set Router Information SR parameters */
+	ospf_router_info_update_sr(true, OspfSR.srgb, 0);
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (sr_prefix_sid,
+       sr_prefix_sid_cmd,
+       "segment-routing prefix A.B.C.D/M index (0-65535)",
```

Comment:
There are a few issues with this command. It's impossible to update the index once the command is typed for the first time (that'd be acceptable only if an error was displayed). Also, this command doesn't show up in the running configuration if you type it after ospfd has started (if it's in the config file then it's processed correctly).

Reply:
The NODE SID must be unique per prefix. See draft. The command here is to allow the configuration of this NODE SID. So, we must check that 1) it is not already used, 2) that the prefix belongs to a loopback interface. What it is feasible is to create several Loopback, with different prefix, and thus, allocate a different index per loopback. I'm also thinking about a similar command at the interface level to simplify that.
After ospfd has started, the command show up only if the corresponding Extended Prefix has been advertised. If not, the index is not registered like in the deferred mode. I will also change that to avoid confusion.